resampler/src/speex.rs

```diff
@@ -1358,3 +1528,86 @@ fn speex_resampler_magic<'a, 'b>(
     *out = &mut value[(out_len * st.out_stride as u32) as usize..];
     out_len
 }
+
+#[cfg(all(
+    any(target_arch = "x86", target_arch = "x86_64"),
+    target_feature = "sse2"
+))]
+#[inline]
+unsafe fn convert_m128_to_m256d(v: __m128) -> __m256d {
+    #[cfg(target_arch = "x86")]
+    use std::arch::x86::*;
+    #[cfg(target_arch = "x86_64")]
+    use std::arch::x86_64::*;
+
+    let v64 = _mm256_castpd128_pd256(_mm_cvtps_pd(v));
+    let v64 = _mm256_insertf128_pd(
+        v64,
+        _mm_cvtps_pd(_mm_shuffle_ps(v, v, 0b00001110)),
+        1,
+    );
+    v64
+}
+
+#[cfg(all(
+    any(target_arch = "x86", target_arch = "x86_64"),
+    target_feature = "sse2"
+))]
+#[inline]
+unsafe fn hsum_m256d(v: __m256d) -> f64 {
+    #[cfg(target_arch = "x86")]
+    use std::arch::x86::*;
+    #[cfg(target_arch = "x86_64")]
+    use std::arch::x86_64::*;
+
+    let low = _mm256_extractf128_pd(v, 0);
+    let high = _mm256_extractf128_pd(v, 1);
+    let low = _mm_add_pd(low, high);
+    let high64 = _mm_unpackhi_pd(low, low);
+    _mm_cvtsd_f64(_mm_add_sd(low, high64))
+}
+
+#[cfg(test)]
+mod tests {
+    #[cfg(target_arch = "x86")]
+    use std::arch::x86::*;
+    #[cfg(target_arch = "x86_64")]
+    use std::arch::x86_64::*;
+
+    #[cfg(all(
+        any(target_arch = "x86", target_arch = "x86_64"),
+        target_feature = "sse2"
+    ))]
+    #[test]
+    fn test_convert_m128_to_m256d() {
+        let input: Vec<f32> = vec![1.0, 2.0, 3.0, 4.0];
+        let mut output: Vec<f64> = vec![0.0; 4];
+
+        unsafe {
+            let v = _mm_loadu_ps(input.as_ptr());
+            _mm256_storeu_pd(
+                output.as_mut_ptr(),
+                super::convert_m128_to_m256d(v),
+            );
+        }
+
+        assert_eq!(output, vec![1.0, 2.0, 3.0, 4.0]);
+    }
+
+    #[cfg(all(
+        any(target_arch = "x86", target_arch = "x86_64"),
+        target_feature = "sse2"
+    ))]
+    #[test]
+    #[allow(clippy::float_cmp)]
```

Comment:
not sure whether this is good idea but im reluctant to introduce tolerance here :man_shrugging:

Reply:
You can do something like [this](https://github.com/rust-av/speexdsp-rs/blob/master/fft/tests/fftwrap.rs#L156) to compare floats.

Reply:
yeah but i mean these are hard coded values and they dont have any error due to conversion :disappointed:

Reply:
Cast them to `usize`? I've seen they don't have the mantissa